packages/worker/src/worker.ts

```diff
@@ -531,25 +535,67 @@ export class Worker {
    *
    * @experimental - this API is considered unstable
    */
-  public static async runReplayHistories(
+  public static async *runReplayHistories(
     options: ReplayWorkerOptions,
     histories: ReplayHistoriesIterable
-  ): Promise<ReplayResults> {
+  ): AsyncIterableIterator<ReplayResult> {
     const [worker, pusher] = await this.constructReplayWorker(options);
     const rt = Runtime.instance();
-    const replayDone = new Subject<void>();
-    const feed = from(histories).pipe(
-      concatMap(async ({ history, workflowId }) => {
-        const validated = this.validateHistory(history);
-        await rt.pushHistory(pusher, workflowId, validated);
-      })
-    );
-    const feeder = lastValueFrom(feed.pipe(takeUntil(replayDone))).finally(() => rt.closeHistoryStream(pusher));
-    const replayer = this.runReplayerToCompletion(worker, { failFast: options.failFast }).finally(() =>
-      replayDone.next()
-    );
-    const [results] = await Promise.all([replayer, feeder]);
-    return results;
+    const evictions = on(worker.evictionsEmitter, 'eviction') as AsyncIterableIterator<[EvictionWithIDs]>;
+    const runPromise = worker.run().then(() => {
+      throw new native.ShutdownError('Worker was shutdown');
+    });
+    void runPromise.catch(() => {
+      // ignore to avoid unhandled rejections
+    });
+    let innerError = undefined;
+    try {
+      try {
+        for await (const { history, workflowId } of histories) {
+          const validated = this.validateHistory(history);
+          await rt.pushHistory(pusher, workflowId, validated);
+          const next = await Promise.race([evictions.next(), runPromise]);
+          if (next.done) {
+            break; // This shouldn't happen, handle just in case
+          }
+          const [{ runId, evictJob }] = next.value;
+          // We replay one workflow at a time so the workflow ID comes from the histories iterable.
+          const error = evictionReasonToReplayError(evictJob, workflowId, runId);
+          yield {
+            workflowId,
+            runId,
+            error,
+          };
+        }
+      } catch (err) {
+        innerError = err;
+      }
+    } finally {
+      try {
+        rt.closeHistoryStream(pusher);
+        worker.shutdown();
+      } catch {
+        // ignore in case worker was already shutdown
+      }
+      try {
+        await runPromise;
+      } catch (err) {
+        /* eslint-disable no-unsafe-finally */
+        if (err instanceof native.ShutdownError) {
+          return;
```

Comment:
Seems to me that we shoud check and rethrow `innerError` in this block

Reply:
Yeah, you're right, thanks.